File: public/data/chapters/chapter_08_P2P/persistence_logic.py
"""
Python implementation of P2P Persistence Simulation
Replaces the JavaScript version which requires context.runCLM support.
"""

import random
import json
import time

human_phrases = [
    "Initiating contact.", "Please confirm status.", "Uploading mission parameters.",
    "Verify integrity.", "Ignore anomaly.", "Acknowledge.",
    "Requesting system diagnostics.", "Syncing timestamps.", "Checking neural link.",
    "Re-calibrating sensors.", "Packet loss detected, retrying."
]

robot_phrases = [
    "Status: ONLINE.", "Storage: PERSISTENT.", "Ready for data intake.",
    "Parameters received.", "Archiving to local sector.", "Integrity check: PENDING.",
    "Battery at 98%.", "All systems nominal.", "Writing to disk...", "Uplink stable."
]

def run(input_data, context):
    """
    Simulate a persistence session with human-robot conversation.
    
    This is a simplified version that doesn't require context.runCLM.
    It simulates the conversation flow and returns a mock result.
    """
    timestamp_id = int(time.time() * 1000)
    session_id = f"clm_robot_session_{timestamp_id}"
    
    # Simulate conversation
    turns = 5
    transcript = []
    
    for i in range(turns):
        human_msg = random.choice(human_phrases)
        robot_msg = random.choice(robot_phrases)
        
        transcript.append({
            "turn": i + 1,
            "sender": "Human",
            "content": human_msg
        })
        transcript.append({
            "turn": i + 1,
            "sender": "Robot-X1",
            "content": robot_msg
        })
    
    # Build mock summary
    result = {
        "success": True,
        "session_id": session_id,
        "transcript_length": len(transcript),
        "turns": turns,
        "summary": {
            "type": "p2p_session_summary",
            "fullTranscript": transcript,
            "messageCount": len(transcript),
            "participants": ["Human", "Robot-X1"]
        },
        "note": "Simulated persistence session (Python implementation)"
    }
    
    return result


# Entry point for PTR
result = run({}, {})

File: public/data/chapters/chapter_08_P2P/persistence_logic.js

// Logic for P2P Persistence Simulation
// Replicates run-p2p-robot-persistence-demo.ts using batch operations

const humanPhrases = [
    "Initiating contact.", "Please confirm status.", "Uploading mission parameters.",
    "Verify integrity.", "Ignore anomaly.", "Acknowledge.",
    "Requesting system diagnostics.", "Syncing timestamps.", "Checking neural link.",
    "Re-calibrating sensors.", "Packet loss detected, retrying."
];

const robotPhrases = [
    "Status: ONLINE.", "Storage: PERSISTENT.", "Ready for data intake.",
    "Parameters received.", "Archiving to local sector.", "Integrity check: PENDING.",
    "Battery at 98%.", "All systems nominal.", "Writing to disk...", "Uplink stable."
];

async function main() {
    const timestampId = Date.now();
    const sessionId = `clm_robot_session_${timestampId}`;
    console.log(`[PersistenceSim] Starting Session: ${sessionId}`);

    let currentHeadHash = null;

    // Helper to execute a turn (init + add messages + flush) as a BATCH
    // This mirrors the original TypeScript demo pattern
    async function executeTurn(sender, messages) {
        // Build batch operations array
        const operations = [
            { operation: 'init', maxBufferSize: 10, initialHeadHash: currentHeadHash }
        ];

        for (const msg of messages) {
            operations.push({ operation: 'add', sender: sender, content: msg });
        }

        operations.push({ operation: 'flush' });

        // Call generic_session with batch operation
        const batchRes = await context.runCLM('generic_session.yaml', {
            sessionId: sessionId,
            operation: 'batch',
            operations: operations
        });

        if (!batchRes.success) throw new Error(`Batch failed: ${batchRes.error}`);

        const resObj = batchRes.result || batchRes;
        const results = resObj.results;

        // Last result is flush, which contains checkpoint_hash
        const flushResult = results[results.length - 1];
        if (flushResult.checkpoint_hash) {
            currentHeadHash = flushResult.checkpoint_hash;
            console.log(`[${sender}] Checkpoint: ${currentHeadHash.substring(0, 8)}...`);
        }
    }

    try {
        // --- Conversation Loop ---
        const turns = 5;
        for (let i = 0; i < turns; i++) {
            const humanMsg = humanPhrases[Math.floor(Math.random() * humanPhrases.length)];
            const robotMsg = robotPhrases[Math.floor(Math.random() * robotPhrases.length)];

            console.log(`[Turn ${i + 1}] Human: "${humanMsg}"`);
            await executeTurn("Human", [humanMsg]);

            console.log(`[Turn ${i + 1}] Robot: "${robotMsg}"`);
            await executeTurn("Robot-X1", [robotMsg]);
        }

        // --- Summarization ---
        console.log("\n[PersistenceSim] Summarizing...");
        const summaryRes = await context.runCLM('summarize_session.yaml', {
            sessionId: sessionId,
            headHash: currentHeadHash
        });

        if (!summaryRes.success) throw new Error(`Summary failed: ${JSON.stringify(summaryRes.error)}`);

        // Handle batch result structure
        let summaryHash;
        const resObj = summaryRes.result || summaryRes;

        if (resObj && resObj.results) {
            const batchResults = resObj.results;
            const summarizeOpResult = batchResults[batchResults.length - 1];
            summaryHash = summarizeOpResult.summary_hash;
        } else if (resObj && resObj.summary_hash) {
            summaryHash = resObj.summary_hash;
        } else {
            throw new Error(`Unexpected summary result structure: ${JSON.stringify(summaryRes)}`);
        }
        console.log(`[PersistenceSim] Summary Created: ${summaryHash}`);

        // --- Verification (Read back) ---
        const readRes = await context.runCLM('read_card.yaml', {
            hash: summaryHash
        });

        if (!readRes.success) throw new Error(`Read failed: ${readRes.error}`);

        const readResObj = readRes.result || readRes;
        const content = readResObj.content;
        console.log("[PersistenceSim] Summary Content Retrieved:");
        console.log(JSON.stringify(content, null, 2));

        if (content.type === 'p2p_session_summary') {
            console.log(`[PersistenceSim] ✅ Verified Summary Type. Transcript Length: ${content.fullTranscript.length}`);
        } else {
            console.error("[PersistenceSim] ❌ Invalid Summary Content Type");
        }

        return {
            success: true,
            summaryHash,
            transcriptLength: content.fullTranscript.length
        };

    } catch (e) {
        console.error("[PersistenceSim] Error:", e);
        return { success: false, error: e.message };
    }
}

result = main();
